### PR TITLE
Optimize the kata qemu binary size

### PR DESCRIPTION
**Pull request by justin-he**

### CODE REVIEW

**Comment by justin-he** [comment]
@kata-containers/builder

File: scripts/configure-hypervisor.sh [hunk 1]
```diff
@@ -354,6 +351,11 @@ generate_qemu_options() {
 		qemu_options+=(size:--disable-dmg)
 		qemu_options+=(size:--disable-parallels)
 		qemu_options+=(size:--disable-vxhs)
+
+		if [[ "${qemu_version_minor}" -ge 1 ]]; then
```
Comment (devimc) [inline scripts/configure-hypervisor.sh]: this won't work for versions >= 4 with a minor equals to 0, for example `4.0.0`.

I think the expression we want here is `[[ "${qemu_version_major}" -ge 4 || ( "${qemu_version_major}" -eq 3  &&  "${qemu_version_minor}" -ge 1 ) ]]`

File: scripts/configure-hypervisor.sh [hunk 2]
```diff
@@ -326,17 +326,14 @@ generate_qemu_options() {
 	# implicitly enabled in Fedora 27).
 	qemu_options+=(size:--disable-linux-aio)
 
-	if [[ "${qemu_version_major}" -ge 4 || ( "${qemu_version_major}" -eq 3  &&  "${qemu_version_minor}" -ge 1 ) ]]; then
+	if [[ "${qemu_version_major}" -ge 4 || ( "${qemu_version_major}" -eq 3  &&  "${qemu_version_minor}" -ge 0 ) ]]; then
```
Comment (devimc) [inline scripts/configure-hypervisor.sh]: thanks, I think now we can simplify this, how about `if [[  "${qemu_version_major}" -ge 3 ]]`?

File: scripts/configure-hypervisor.sh [hunk 3]
```diff
@@ -283,13 +283,11 @@ generate_qemu_options() {
 		qemu_options+=(misc:--static)
 	fi
 
-	# Disable debug and "-uuid ..." is always passed to the qemu binary so not required.
```
Comment (devimc) [inline scripts/configure-hypervisor.sh]: any idea about in which release this option was deprecated?

File: scripts/configure-hypervisor.sh [hunk 4]
```diff
@@ -318,7 +318,7 @@ generate_qemu_options() {
 
 	# Disable XEN driver
 	case "$arch" in
-	aarch64) ;;
+	aarch64) qemu_options+=(size:--disable-xen) ;;
```
Comment (devimc) [inline scripts/configure-hypervisor.sh]: since this is the section where XEN is disabled, how about disable it for all architectures?

Comment (justin-he) [reply scripts/configure-hypervisor.sh]: sure

File: scripts/configure-hypervisor.sh (hunk 3 above)
Comment (justin-he) [reply scripts/configure-hypervisor.sh]: @devimc Please refer to commit 315d318 of qemu
git describe --contains 315d318
v2.8.0-rc0~121^2~14

File: scripts/configure-hypervisor.sh (hunk 1 above)
Comment (justin-he) [reply scripts/configure-hypervisor.sh]: Yes, would be better

File: scripts/configure-hypervisor.sh (hunk 2 above)
Comment (justin-he) [reply scripts/configure-hypervisor.sh]: sure

**Comment by Pennyzct** [comment]
Hi~ guys 
obs ci failed on the following error:
```
make: *** No rule to make target '/tmp/build-image-tmp.XbA3znzVhz/initrd-image/.*default-initrd-name_rootfs.done', needed by '/tmp/build-image-tmp.XbA3znzVhz/osbuilder/kata-containers-initrd.img'.  Stop.
```
It looks to me that yq here failed on traversing anchor `*default-initrd-name`. @chavafg has fixed this issue in the latest PR https://github.com/kata-containers/tests/pull/2297.

**Comment by jcvenegas** [comment]
@Pennyzct  thanks for the heads up lets run it again to see the  logs

**Comment by egernst** [comment]
Looks good. Can you update the commit messages to reflect this isn’t aarch64 specific?

**Comment by justin-he** [comment]
> Looks good. Can you update the commit messages to reflect this isn’t aarch64 specific?

Yes, sure

**Comment by jcvenegas** [comment]
Waiting for https://github.com/kata-containers/packaging/pull/935 to unblock CI

**Comment by GabyCT** [comment]
@justin-he could you please resolve the conflicts? thanks

**Comment by justin-he** [comment]
> @justin-he could you please resolve the conflicts? thanks

Hi @GabyCT , resolved the conflicts with commit 
commit 83a69de4a24d50338b8f9a0962a21224612bf4ce
Author: Julio Montes <julio.montes@intel.com>
Date:   Tue Mar 3 17:59:31 2020 +0000

    scripts: enable libpmem only for x86_64

**Comment by amshinde** [comment]
@justin-he Looks like the optimizations are no longer arch specific. Can you update the commit message to reflect this.

**Comment by justin-he** [comment]
> @justin-he Looks like the optimizations are no longer arch specific. Can you update the commit message to reflect this.

Ok, I changed the patch 2,3,5 commit msg, described that they are for generic arches. But I keep the aarch64 in patch 1,4 because I thought --disable-xen and --pie had been enabled on others , only aarch64 needs these 2 patches.

File: scripts/configure-hypervisor.sh [hunk 5]
```diff
@@ -317,12 +317,7 @@ generate_qemu_options() {
 	qemu_options+=(size:--disable-tools)
 
 	# Disable XEN driver
-	case "$arch" in
-	aarch64) ;;
-	x86_64) qemu_options+=(size:--disable-xen) ;;
-	ppc64le) qemu_options+=(size:--disable-xen) ;;
-	s390x) qemu_options+=(size:--disable-xen) ;;
-	esac
+	qemu_options+=(size:--disable-xen)
```
Comment (devimc) [inline scripts/configure-hypervisor.sh]: thanks, could you please update the commit message? :smile_cat:

Comment (justin-he) [reply scripts/configure-hypervisor.sh]: sure， updated

**Comment by amshinde** [comment]
@justin-he Changes look good. Do you have details about how much reduction in binary size are seen with your changes?

**Comment by justin-he** [comment]
> @justin-he Changes look good. Do you have details about how much reduction in binary size are seen with your changes?

Sure, the details are as follows:
```
root@server# ll aarch64-softmmu.old/qemu-system-aarch64 
-rwxr-xr-x 1 root root 71977064 Mar 11 16:22 aarch64-softmmu.old/qemu-system-aarch64*
root@server# ll aarch64-softmmu.new/qemu-system-aarch64    
-rwxr-xr-x 1 root root 70795448 Mar 11 16:39 aarch64-softmmu.new/qemu-system-aarch64*
```
It reduced about 1MBit

**Comment by amshinde** [comment]
> It reduced about 1MBit

Thats great! Thanks for the change @justin-he

**Pull request merged**